File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

sys.setrecursionlimit(20000)

def main():
    nums = list(map(int, sys.stdin.read().split()))
    def postorder(start, end):
        if start >= end: return
        root = nums[start]
        low, high= start+1, end
        while low < high:
            mid = (low+high)>>1
            if nums[mid] <= root: low = mid+1
            else: high = mid
        postorder(start+1, low)
        postorder(low, end)
        print(root)
    postorder(0, len(nums))
    return
if __name__ == '__main__':
    main()

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

vector<int> nums;

void postorder(int start, int end){
    if(start >= end) return;
    int root = nums[start];
    int low = start+1, high = end;
    while(low < high){
        int mid = (low+high)>>1;
        if(nums[mid] <= root) low = mid+1;
        else high = mid;
    }
    postorder(start+1, low);
    postorder(low, end);
    cout << root << '\n';
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int x;
    while(cin >> x) nums.push_back(x);
    postorder(0, nums.size());
    return 0;
} 